### PR TITLE
Ruby enumerables

### PR DESCRIPTION
**Pull request by vichuge**
I learn about the Enumerable module that gets mixed into the Array and Hash classes (among others) and I was provided with lots of handy iterator methods. To prove that there's no magic to it, I'm going to rebuild those methods.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,38 @@
-# ruby_enumerables
-You learned about the Enumerable module that gets mixed into the Array and Hash classes (among others) and provides you with lots of handy iterator methods. To prove that there's no magic to it, you're going to rebuild those methods.
+![](https://img.shields.io/badge/Microverse-blueviolet)
+
+# ruby_enumerables
+
+> You learned about the Enumerable module that gets mixed into the Array and Hash classes (among others) and provides you with lots of handy iterator methods. To prove that there's no magic to it, you're going to rebuild those methods.
+
+**This project was buid with ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]**
```
Comment (kitpao) [inline README.md]: - [x] **OPTIONAL** You may want to correct the typo buid -> build

File: house.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  def my_select
+    arr = to_a
+    new_a = []
+    (0..arr.length - 1).each do |i|
+      next unless yield(arr[i])
+
+      new_a << arr[i]
+    end
+    p new_a
```
Comment (kitpao) [inline house.rb]: - [x] Please, do not use `p`, `puts`, or `print` in your methods, remember that the original methods don't show anything into the console unless it is explicitly required

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @vichuge 


**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

- For next time, please be more descriptive in your commits, you made only 1 commit for all the methods.

- [x] Do not forget to add the linters for the project, you can find the [instructions here](https://github.com/microverseinc/linters-config/tree/master/ruby), please add the necessary files and check that you get a message similar to this one in your pull request:
image

### Methods:
  - [ ] Please notice that the original methods take into account what happens when a block is not given (They return an enumerator object). This applies to my_each, my_each_with_index, my_select, my_map **Hint:** you can check [this documentation for more](https://apidock.com/ruby/Object/to_enum).
  **Tip:** You can see the complete [description of all the methods here](https://ruby-doc.org/core-2.6.5/Enumerable.html), please check it out as a guide if you feel lost

#### #my_each
  - [x] Please check and correct the difference between original and your methods:
  ```
  p (1..3).each{ |x| x }               # 1..3
  p (1..3).my_each{ |x| x }             # 0..2
  ```
  Notice that the original returns the original range, and yours is modifying it. **Hint:** Don't forget to return something at the end
#### #my_each_with_index
  - [x] Please check and correct the difference between original and your methods:
  ```
  p (1..3).each_with_index{ |x| x }               # 1..3
  p (1..3).my_each_with_index{ |x| x }             # 0..2
  ```
  Same as before, notice that the original returns the original range, and yours is modifying it. **Hint:** Don't forget to return something at the end

#### #my_all?
- [x] Please check and correct the difference between original and your method:
  ```
  p [1,false,3].all?               # false
  p [1,false,3].my_all?            # true
  ```
  Notice that the original returns True and yours is returning False. **Hint:** You are only returning false when `arr[i].nil?`

- [x] Please check and correct the difference between original and your method:
  ```
  p [1,2,3].all?               # true
  p [1,2,3].my_all?            # false
  ```
  Notice that the original returns True and yours is returning False. **Hint:** You are only returning false when `arr[i].nil?`

  - [x] Please notice that your methods should be able to take Classes, Regexp, and Patterns as parameters, 
**Hint:** Investigate the "===" operator, it can be very useful for this
  ```
    p ['d', 'dd'].all?(/d/)              # true
    p ['d', 'dd'].my_all?(/d/)           # TypeError
    p ['e', 'er'].all?("e")              # false
    p ['e', 'er'].my_all?("e")           # TypeError
  ```
 - [ ] **OPTIONAL** Advanced hint: you can use my_all? method inside the definition of my_all?. example: if no block is given, then do `arr.my_all?{# I give my own block}`

#### #my_any?
- [x] Please check and correct the difference between original and your method:
  ```
  p [false,false,false].any?               # false
  p [false,false,false].my_any?            # true
  ```
  Notice that the original returns False and yours is returning True

  - [x] Please notice that your methods should be able to take Classes, Regexp, and Patterns as parameters, 
**Hint:** Investigate the "===" operator, it can be very useful for this
  ```
    p ['d', 'dd'].any?(/d/)              # true
    p ['d', 'dd'].my_any?(/d/)           # TypeError
    p ['ex', 'er'].any?("e")              # false
    p ['ex', 'er'].my_any?("e")           # TypeError
  ```
 - [ ] **OPTIONAL** Advanced hint: you can use my_any? method inside the definition of my_any?. example: if no block is given, then do `my_any?{# I give my own block}`

#### #my_none?
  - [x] Please notice that your methods should be able to take Classes, Regexp, and Patterns as parameters, 
**Hint:** Investigate the "===" operator, it can be very useful for this
  ```
    p ['q', 'qq'].none?(/d/)              # true
    p ['d', 'dd'].my_none?(/d/)           # TypeError
    p ['e', 'er'].none?("e")              # false
    p ['e', 'er'].my_none?("e")           # TypeError
  ```
 - [ ] **OPTIONAL** Advanced hint: you can use my_none? method inside the definition of my_none?. example: if no block is given, then do `my_none?{# I give my own block}`

#### #my_count
  - [x] `p [3,2,1,0,1,1,1].my_count{|x| x >= 2}` should return 2 because there are 2 elements that have that condition in the block. Your method returns 6. **Hint:** Instead of using `(0..arr.length).my_each`, why you don't pass the array directly? `arr.my_each`. because 0..arr.length is giving you numbers instead of the items

#### #my_map
  - [x] Your code doesn't prioritize the proc over the block as required in Odin, yours is giving priority to the block
```
  my_proc = proc { |num| num > 10 }
  p [5,6,7,8,9,10,11,12,13].my_map(my_proc) { |num| num < 10 } # --> should be [false, false, false, true, true]
```
image

#### #my_inject

  - [x] Your method should be able to manage symbols for operations (:+, :*, etc) [Check the examples in the documentation here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)
  - [x] `[1,2,3].inject(2, :*)` should be 12 (2 * 1 * 2 *3), your method doesn't accept 2 arguments
  - [x] `["rod", "blade", "qq"].my_inject{ |memo, word| memo.length > word.length ? memo : word }` should be "blade" because it is the largest string. Your method breaks with this input
  - [x]  `[2,3,4].my_inject(:+)` should return 9 -> 2+3+4=9. Your code doesn't know what to do with the symbol.

#### #multiply_els
  - [x] `multiply_els [2, 4, 5]` should return 40, your method is undefined for an array **Hint:** take that method out of the module and pass the array as an argument

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
### Hello @vichuge, you are awesomely well with the level of progress I'm seeing.

Let's address a few things, we are all learning here, and do feel free to reach out to any of the TSEs on slack if you feel stuck on a particular task, be it setting up the GitHub flow or Rubocop issues, we will be more than happy to help. Your level is already high, you can do it 😄 

- [x] Please kindly add the remarks about "thinking is a bad idea to ask...." to comment and replace the Pull Request Title with a descriptive one that reflects what you did on this PR. Thank you 🙏 

- [x] Again do not forget to add the linters for the project, you can find the i[nstructions here](https://github.com/microverseinc/linters-config/tree/master/ruby). Please feel free to contact me if you need any assistance. 
Screen Shot 2020-12-04 at 12 11 02 AM

#### You are almost there, only 5 test cases are failing, kindly fix them are we will be done

- [x] Your `#my_any?` should return false if the block ever returns a value false or nil
```ruby

false_block = proc { |num| num > 9 }
array = Array.new(100) { rand(0...9) }
array.my_any?(&false_block)    # => false

```

- [x] Your `#my_any?` when no block or argument is given should return false if at least one of the collection is not true
```ruby
[nil, false, nil, false].my_any?
```

- [x] Your `#my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
You can compare these two to check the validity of your code, the results should be the same comparing against the original `#inject` method.
```ruby
        actual = range.my_inject { |prod, n| prod * n }
        expected = range.inject { |prod, n| prod * n }

```

- [x] Your `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
You can compare the results of the two values to check the correctness of your method
```ruby
range = Range.new(5, 50)
expect(range.my_inject(:*)).to eq range.inject(:*)
```

- [x] Move `#multiply_els` outside the **module** and it should accept an array as an argument and multiplies all the elements of the array together using #my_inject


### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: house.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    arr.length.times do |i|
+      return true if block_given? && yield(arr[i])
+      return true if !param.nil? && arr[i].is_a?(param)
+      return false && exist_falses = true if arr[i] == false
+    end
+    return false if arr.empty?
+    return true if exist_falses == false && !block_given? && param.nil?
+
+    false
+  end
```
Comment (ghost) [inline house.rb]: Fix the following failing test case for `my_any?` method.

` %w[ant bear cat].my_any?(/d/)`

File: house.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      end
+    else
+      sta = arr[0]
+      arr.length.times do |i|
+        sta = yield(sta, arr[i])
+      end
+    end
+    sta
+  end
+end
```
Comment (ghost) [inline house.rb]: The following test cases are failing for `my_inject` method.
```
(5..10).my_inject(:-)
(5..10).my_inject(:*)
```

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

## Hello team 😄,

**Great job implementing the requested changes 💪. Just a few fixes needed and you should be good.**

1). **Fix the failing linter checks.**
2). **You still have rubocop offenses being detected.Create a `.rubocop.yml` file for your `rubocop` errors and that should help you with this. Check this [guideline](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml) from Microverse. Hint: At the top of the file page, where it reads `Exclude`, add the following files so they are excluded from rubocop check,**

```
AllCops:  
  Exclude:
 - "db/**/*"
- "bin/*"
 - "config/**/*"    
- "Guardfile"   
 - "Rakefile"   
 - "node_modules/**/*"
```
3). **Check for more in the comments below.**

### Happy coding 🥂

**PS: You can reach me on slack (Rose Wanjohi).**

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      return true if block_given? && yield(arr[i])
+      return true if param.is_a?(Class) && arr[i].is_a?(param)
+      return true if param.instance_of?(Regexp) && arr[i].to_s.match(param.to_s)
+      return false && exist_falses = true if arr[i] == false
+    end
+    return false if arr.empty?
+    return true if exist_falses == false && !block_given? && param.nil?
+
+    false
+  end
```
Comment (elmejdki) [inline main.rb]: ## my_any?:

- [x] when no block or argument is given returns true if at least one of the collection is not false or nil:

```ruby

    true_array = [1, true, 'hi', []] 
    true_array.my_any? === true_array.any? => false

```

- [ ] when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern:

```ruby

   words = %w[cats door rod blade] 
   words.my_any?('cat') === words.any?('cat') => false

```

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    elsif arr[0].is_a?(Integer)
+      arr.length.times do |i|
+        sta = yield(sta, arr[i])
+      end
+    else
+      sta = arr[0]
+      arr.length.times do |i|
+        sta = yield(sta, arr[i])
+      end
+    end
```
Comment (elmejdki) [inline main.rb]: ## my_inject:

- [ ] when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range:

```ruby

    range = (1..99)
    range.my_inject(2, :*) === range.inject(2, :*) => false

```

- [ ] when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range:

```ruby

    range = (1..99)
    actual = range.my_inject { |prod, n| prod * n }
    expected = range.inject { |prod, n| prod * n }
    actual === expected => false

```

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    sta
+  end
+end
+
+def multiply_els(elements)
+  elements.my_inject { |one, two| one * two }
+end
```
Comment (elmejdki) [inline main.rb:133]: ## multiply_els:

- [ ] accepts an array as an argument and multiplies all the elements of the array together using #my_inject:

```ruby

      actual = multiply_els [2, 4, 5]
      expect(actual).to eq 40

```

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

Great job so far @vichuge :+1: :100: :clap:, but you still have some minor changes :hammer: :wrench: to take care of.

## Required changes:

Please check the comments below.
### :point_down: :point_down: :point_down: :point_down: :point_down: :point_down: :point_down: :point_down: 

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_If you have any questions don't hesitate to contact me on slack._

File: main.rb (hunk 6 above)
Comment (vichuge) [reply main.rb]: image

Now I'm checking the method and the result is equal to the original.

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:
### Hi @vichuge ! 👋 👋 

### You have done well so far and you're almost done! :clap: However, there are changes required for you to make for this project to be approved. Kindly work on them and submit them for another code review.

**Recommended**

- [x]  `#my_any?` when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern.

- [x] `#my_inject` when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block

- [x] ` #my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Review by Temmarie — APPROVED** [review]
## Status: Approved :heavy_check_mark:

### Hi @vichuge ! 👋 👋 

#### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: :tada:

#### You may merge, good luck with your next projects! :rocket:
#### Happy Coding! :+1:

__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Pull request merged**